(data) Update the Ahrefs success story and add a new story about the full stack ocaml web solution (#2863)
ocaml/ocaml.org: data/success_stories/ahrefs.md

---

> title: Marketing Analysis Tools Powered by Big Data

Reviewer: ```suggestion
title: Marketing Intelligence Tools Powered by Big Data
```

logo: success-stories/ahrefs.svg
card_logo: success-stories/white/ahrefs.svg
background: /success-stories/ahrefs-bg.jpg
theme: blue

> synopsis: "Ahrefs leverags OCaml to achieve massive scalability, processing billions of daily frontend and backend requests, running a peta-byte scale web crawler, while maintaining a lean, efficient team."

Reviewer: ```suggestion
synopsis: "Ahrefs leverages OCaml to achieve massive scalability, processing billions of daily requests, running a peta-byte scale web crawler, while maintaining a lean, efficient team."
```

> peta-byte scale web crawler

might be a bit abstract? 3rd biggest crawler in the world, after google, before bing

https://radar.cloudflare.com/traffic/verified-bots
![image](https://github.com/user-attachments/assets/a9aa83b0-0439-4412-a045-785c6c430c0f)

Author: yeah, putting it in relation to the rest of the market is awesome!

url: https://ahrefs.com/
priority: 2
why_ocaml_reasons:
- Type Safety
- Unified Technology Stack
- Expressiveness and Reliability
- Integration with JavaScript Ecosystem
- Customizability

> - Scalability

Reviewer: ```suggestion
- Type Safety
- Expressiveness and Reliability
- Unified Technology Stack
- Integration with JavaScript Ecosystem
- Customizability
- Scalability
- Performance
```

Type Safety and Reliability are kind of redundant?

Unified Technology Stack and Integration with JavaScript Ecosystem potentially too? I've grouped them because they are related

---

## Challenge

Ahrefs has always been committed to operational efficiency and self-reliance, even as it grew to become a leader in the SEO industry. Unlike many VC-funded companies, Ahrefs has chosen to prioritize maintaining a small, focused team while achieving exceptional output. However, this approach came with its own set of challenges.

> In 2017, the state of technology presented significant limitations. JavaScript tooling was far from ideal, offering poor editor integration and limited safety features, which hindered developer productivity. Similarly, the PHP and jQuery stack being used lacked the structure and robustness of a modern frontend framework. This resulted in inefficiencies and created communication barriers between frontend and backend teams.

Reviewer: ```suggestion
```

> offering poor editor integration

ocaml/bucklescript is/was even poorer, I don't think it's a good argument. It's not what drove our decisions.

I'm not convinced that this whole paragraph is necessary. Not point "attacking" other ecosystem. Especially when they happen to be much more mature than ocaml

Another challenge was scaling operations while keeping the team small. Ahrefs needed a cohesive and type-safe technology stack that worked seamlessly across the frontend and backend, ensuring confidence in the codebase and maintaining a high standard of output without the need to hire excessively.

> Finally, the lack of a direct connection between the frontend and backend often required backend engineers to manually create APIs, leading to delays and inefficiencies in delivering features. To address these challenges, Ahrefs sought a solution that could empower their lean team to work more effectively, with minimal friction and strong collaboration between all parts of the stack.

Reviewer: ```suggestion
One major challenge was maintaining high-quality standards across the codebase. Ahrefs needed a cohesive and type-safe technology stack that worked seamlessly across the frontend and backend, enabling developers to write reliable code with confidence and ensuring a consistent, high-quality output.

Another challenge was scaling operations. The lack of a integration between the different bricks often required teams to wait on one another, needing collaboration to create APIs, leading to delays and inefficiencies in delivering features. The team was looking for a unified stack, empowering all engineers to contribute to any part of the stack, reducing bottlenecks, increasing ownership and business knowledge, while maintaining a high trust in the final result.
```

Feels like quality and throughput should be two different points rather than intertwined.

> needing collaboration to create APIs

I don't really like this bit, mentioning APIs only is a bit reductive, but I don't have a much better idea on how to word it for now.

## Result

Adopting OCaml was a transformative decision for Ahrefs, enabling the company to overcome its challenges and significantly improve operations. By transitioning the entire frontend team from PHP and jQuery to OCaml, supported by Melange and React, Ahrefs achieved a full-stack evolution.

One of the most impactful outcomes was the ability to share data types between the frontend and backend. This streamlined development processes, reduced friction, and allowed engineers to contribute meaningfully across the entire stack. With this new technology foundation, Ahrefs not only maintained but significantly scaled its operations.

> Today, Ahrefs’ systems handle **5 billion frontend requests per day** and process a staggering **500 billion backend requests per day**. The codebase powering this scale has grown to **1.5 million lines of OCaml code**, underscoring the robustness and scalability of their technology stack. These advancements have helped Ahrefs become a trusted industry leader, relied upon by **44% of Fortune 500 companies**.

Reviewer: > Today, Ahrefs’ systems handle **5 billion frontend requests per day** and process a staggering **500 billion backend requests per day**.

it's actually more ~50M frontend requests. And billions but exact number unknown of internal requests. I gave some incorrect numbers during our call.

Author: changed to just mention millions vs billions for frontend / backend

For more up-to-date statistics, see [Ahrefs' statistics and metrics page](https://ahrefs.com/big-data).

## Why OCaml

Ahrefs chose OCaml as the backbone of its technology stack because of its unique ability to combine expressiveness and reliability. The language allowed developers to write concise, efficient code that remained stable over the long term, with systems often running for years without surprises.

One of the key reasons for selecting OCaml was its versatility in enabling a unified technology stack. The ability to use shared types between frontend and backend eliminated redundant work, improved collaboration, and ensured consistency across the board. Additionally, OCaml’s robust type system provided engineers with a high degree of confidence in their code, significantly reducing errors and simplifying maintenance.

> OCaml’s ecosystem, especially with the introduction of Melange (formerly BuckleScript), allowed Ahrefs to bridge the gap between OCaml’s strong typing and the flexibility of the JavaScript ecosystem. Initially adopted for backend crawling, OCaml’s utility expanded to the frontend when Melange made it possible to compile OCaml to JavaScript. This pivotal decision empowered Ahrefs to create tools that made their vast datasets more accessible and actionable.

Reviewer: ```suggestion
OCaml’s ecosystem, especially with the introduction of Melange (formerly BuckleScript), allowed Ahrefs to bridge the gap between OCaml’s strong typing and the flexibility of the JavaScript ecosystem. Initially adopted for crawling and data processing, OCaml’s utility expanded to the frontend when Melange made it possible to compile OCaml to JavaScript. This decision empowered Ahrefs to create tools that made their vast datasets more accessible and actionable.
```

The unified OCaml stack also fosters a more collaborative environment. By allowing all engineers to contribute across domains, Ahrefs broke down silos and improved overall productivity. The result was a streamlined development process that enabled the company to maintain high standards of output while keeping its team small and focused.

## Lessons Learned

Ahrefs’ journey with OCaml provides valuable insights for other companies considering a similar path:

1. **Embrace Simplicity:** OCaml’s type system is intuitive yet powerful, enabling developers to build expressive and reliable applications without unnecessary complexity.
2. **Invest in Learning:** Training the team in OCaml is essential but rewarding. Tools like the Melange playground can make the onboarding process smoother and more approachable.
3. **Build Tailored Solutions:** By leveraging OCaml’s expressiveness, Ahrefs avoided generic frameworks in favor of custom-built solutions that addressed their unique needs. While this approach requires more ownership, it results in greater efficiency and precision.

> 4. **Anticipate Challenges with Bindings:** Writing bindings can be complex, but resources from the Melange community simplify this process.

Reviewer: ```suggestion
4. **Anticipate Challenges with Bindings:** Writing bindings can be complex, but resources from the community simplify this process.
```

I don't understand how anticipation plays a role here.

Removing melange because bindings to C/C++ are equally critical to us

5. **Adopt Gradually:** Starting with small contributions to libraries or limited integrations can help teams build confidence before transitioning fully to OCaml.

## Open Source

Ahrefs is proud to support the OCaml ecosystem by contributing tools and libraries that benefit the broader community. These contributions include:

- **[Melange Recharts](https://github.com/ahrefs/melange-recharts):** Recharts bindings for Melange.

> - **[Ahrefs DevKit](https://github.com/ahrefs/devkit):** A suite of internal tools and utilities.

Reviewer: don't think that "internal" is the right keyword here. devkit is a pretty general toolkit for applications, it's not really tied to anything internal to ahrefs
